Applications/Utils/ModelPreparation/PartitionMesh/PartitionMesh.cpp

```diff
@@ -0,0 +1,127 @@
+/*!
+  \file PartitionMesh.cpp
+  \date   2016.05
+
+  \brief  A tool for mesh partitioning.
+
+  \copyright
+  Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+             Distributed under a Modified BSD License.
+               See accompanying file LICENSE.txt or
+               http://www.opengeosys.org/project/license
+
+*/
+
+#include "tclap/CmdLine.h"
+
+#ifdef BUILD_WITH_METIS
+extern "C" {
+#include "metis_main.h"
+}
+#endif
+
+#include "Applications/ApplicationsLib/LogogSetup.h"
+#include "BaseLib/FileTools.h"
+
+#include "MeshLib/IO/readMeshFromFile.h"
+#include "MeshLib/IO/VtkIO/VtuInterface.h"
+
+#include "BaseLib/CPUTime.h"
+#include "BaseLib/RunTime.h"
+
+#include "MeshPartitioning.h"
+
+int main (int argc, char* argv[])
+{
+    ApplicationsLib::LogogSetup logog_setup;
+
+    std::string m_str = "Partition a mesh for parallel computing."
+                        "The tasks of this tool are in twofold:\n"
+                        "1. Convert mesh file to the input file of the partitioning tool,\n"
+                        "2. Partition a mesh using the partitioning tool,\n"
+                        "\tcreate the mesh data of each partition,\n"
+                        "\trenumber the node indicies of each partition,\n"
+                        "\tand output the results for parallel computing.";
+
+    TCLAP::CmdLine cmd(m_str, ' ', "0.1");
+    TCLAP::ValueArg<std::string> mesh_input("i", "mesh-input-file",
+                                            "the name of the file containing the input mesh", true,
+                                            "", "file name of input mesh");
+    cmd.add(mesh_input);
+    TCLAP::ValueArg<int> nparts("n", "np", "the number of partitions", false, 2, "integer");
+    cmd.add(nparts);
+
+    TCLAP::SwitchArg ogs2metis_flag("1", "ogs2metis",
+                                    "Indicator to convert the ogs mesh file to METIS input file", cmd, false);
+
+    TCLAP::SwitchArg asci_flag("a","asci","Enable ASCII output.", false);
+    cmd.add(asci_flag);
+    TCLAP::SwitchArg elem_part_flag("e","element_wise","Enable element wise partitioing.", false);
+    cmd.add(elem_part_flag);
+
+    cmd.parse(argc, argv);
+
+    BaseLib::RunTime run_timer;
+    run_timer.start();
+    BaseLib::CPUTime CPU_timer;
+    CPU_timer.start();
+
+    const std::string ifile_name = mesh_input.getValue();
+    const std::string file_name_base = BaseLib::dropFileExtension(ifile_name);
+    MeshLib::MeshPartitioning* mesh = static_cast<MeshLib::MeshPartitioning*>
```

Comment:
That's tricky. I don't like the static cast here. The only way around it that comes to my mind now is, that you don't extend `Mesh`, but rather store a reference to a `Mesh` inside `MeshPartitioning`.
Then you can store the `Mesh` in a `uniqure_ptr` and construct the `MeshPartitioning` from it without duplicating memory.
Furthermore: Currently you don't delete `mesh`, so it will leak memory, I guess. (Sure, that's not a big issue since the program ends anyway)